katdal/chunkstore_s3.py

```diff
@@ -338,16 +359,35 @@ def create_array(self, array_name):
             with self._request(None, 'PUT', policy_url, data=json.dumps(policy)):
                 pass
 
+    @classmethod
+    def _numpy_header(cls, chunk):
+        fp = io.BytesIO()
+        header_fields = np.lib.format.header_data_from_array_1_0(chunk)
+        np.lib.format.write_array_header_1_0(fp, header_fields)
+        return fp.getvalue()
+
     def put_chunk(self, array_name, slices, chunk):
         """See the docstring of :meth:`ChunkStore.put_chunk`."""
         chunk_name, _ = self.chunk_metadata(array_name, slices, chunk=chunk)
         url = self._chunk_url(chunk_name)
-        fp = io.BytesIO()
-        np.lib.format.write_array(fp, chunk, allow_pickle=False)
-        md5 = base64.b64encode(hashlib.md5(fp.getvalue()).digest())
-        fp.seek(0)
+        # Note: don't use ascontiguousarray as it turns 0D into 1D.
+        # See https://github.com/numpy/numpy/issues/5300
+        chunk = np.asarray(chunk, order='C')
+        # Need an array of bytes to so that _Multipart.len gives the right value
+        chunk_view = memoryview(np.ravel(chunk).view(np.uint8))
```

Comment:
Why do you need `ravel` and a `uint8` view? I would have thought that `chunk_view = memoryview(chunk)` is sufficient, given that chunk satisfies the buffer protocol and its buffer is already C-contiguous.

Reply:
The problem is that len(memoryview(chunk)) returns the number of items, not the number of bytes, which would break `_Multipart` since it uses `len` to get the number of bytes. The ravel is probably not necessary, I just wasn't sure what `len` would return for a multi-dimensional array.

Thinking about it more though, I think it might be better to handle this in `_Multipart` itself, so that you just put the chunk directly into `_Multipart`, and `_Multipart` takes care of wrapping things into memory views and figuring out their lengths.

Reply:
Done as described above.